### PR TITLE
Adding Default XmlnsDefinition

### LINKED ISSUE #345: [Enhancement | Discussion] Look into using XmlnsDefinition
@dansiegel proposed to use `XmlnsDefinition` to make it easier to consume this library. Thoughts? Pros? Cons?

### PR DESCRIPTION
**Pull request by dansiegel**
### Description of Change ###

Adding XmlnsDefinitions for the Xamarin Community Toolkit. This utilizes a namespace based on what we see from Xamarin.Forms `http://xamarin.com/schemas/{year}/{product}`. The sample app has been updated to use the new `http://xamarin.com/schemas/2020/toolkit` namespace consistently using the `xct` prefix. 

**Q.** Why is the XmlnsDefinition Attribute excluded from .NET Standard 1.0?
**A.** Xamarin.Forms makes this API internal for netstandard1.0 due to an issue with Reflection methods that are not available. For those impacted, they can continue to use a traditional explicit xmlns for each clr namespace they require, or they can simply update to netstandard2.0
image
*from the Xamarin.Forms source*

### Issue ###

- closes #345 

### API Changes ###

none

### Behavioral Changes ###

When consuming the Xamarin Community Toolkit you will no longer need to declare a unique xmlns for each CLR namespace you're consuming in XAML (unless you're oddly using netstandard1.0)

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [x] Rebased on top of main at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

**Review by sthewissen — APPROVED** [review]

**Comment by sthewissen** [comment]
@jfversluis @AndreiMisiukevich Do we want to merge this in? Not sure if more debating on taking on the Forms namespace is needed?

**Comment by dansiegel** [comment]
@sthewissen well I wouldn't say merge this yet... really we need to decide whether or not we're good with using the Forms namespace... if @jfversluis & @AndreiMisiukevich agree, then I'll update the sample app to use the namespace

**Comment by AndreiMisiukevich** [comment]
I like this contribution :) But I am not sure about the exact namespace. @jfversluis we need your input here

**Review by AndreiMisiukevich — DISMISSED** [review]

**Comment by jfversluis** [comment]
I can see it go both ways. I certainly see how this makes it easier, especially with people transitioning controls like the Expander, MediaElement, etc. but might indeed also cause confusion as to where issues need to go.

This means you can do both, right? So you can still do both Xamarin.CommunityToolkit... but you will also see the things pop up in your IntelliSense for Xamarin.Forms?

I guess I am kind of onboard with, this. How do Forms members feel about this? @Redth @jsuarezruiz @mattleibow @PureWeen

**Comment by Redth** [comment]
We should not share the same namespace as Xamarin.Forms as it would be a confusing and misleading.  Let's pick a reasonable namespace (ie: `Xamarin.Forms.Toolkit`) and declare a `XmlnsPrefixAttribute` so that intellisense can help sort out missing prefixes.

**Review by Redth — CHANGES_REQUESTED** [review]
We need a unique namespace prefix and not to reuse the Xamarin.Forms one.

**Comment by dansiegel** [comment]
Given the feedback from @Redth... there are two alternatives that I'm thinking:

- Just go stupid simple: `http://xamarincommunitytoolkit.com`
- Follow the default XF schema (`http://xamarin.com/schemas/2014/forms`): `http://xamarin.com/schemas/2020/toolkit`

Keep in mind that regardless of what we ultimately choose here the XAML intellisense will show you the full domain so it isn't like someone would have to remember what the exact uri is. Anyone have any thoughts?

**Comment by AndreiMisiukevich** [comment]
http://xamarin.com/schemas/2020/toolkit -- looks good
Any thoughts?

**Comment by AndreiMisiukevich** [comment]
> http://xamarin.com/schemas/2020/toolkit -- looks good
> Any thoughts?

Or even ```https://github.com/xamarin/XamarinCommunityToolkit```

**Comment by jfversluis** [comment]
Let's go with this one: http://xamarin.com/schemas/2020/toolkit

Just so we are inline with the rest and it looks super interesting 🤓

File: XamarinCommunityToolkit/AssemblyInfo/AssemblyInfo.shared.cs [hunk 1]
```diff
@@ -1,3 +1,12 @@
...
-[assembly: InternalsVisibleTo("Xamarin.CommunityToolkit.UnitTests")]
\ No newline at end of file
+[assembly: InternalsVisibleTo("Xamarin.CommunityToolkit.UnitTests")]
+
+#if !NETSTANDARD1_0
+[assembly: XmlnsDefinition("http://xamarin.com/schemas/2014/forms", "Xamarin.CommunityToolkit.Behaviors")]
+[assembly: XmlnsDefinition("http://xamarin.com/schemas/2014/forms", "Xamarin.CommunityToolkit.Converters")]
+[assembly: XmlnsDefinition("http://xamarin.com/schemas/2014/forms", "Xamarin.CommunityToolkit.Effects")]
+[assembly: XmlnsDefinition("http://xamarin.com/schemas/2014/forms", "Xamarin.CommunityToolkit.Extensions")]
+[assembly: XmlnsDefinition("http://xamarin.com/schemas/2014/forms", "Xamarin.CommunityToolkit.UI.Views")]
```
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/AssemblyInfo/AssemblyInfo.shared.cs:11]:
```suggestion
[assembly: XmlnsDefinition("http://xamarin.com/schemas/2020/toolkit", "Xamarin.CommunityToolkit.Behaviors")]
[assembly: XmlnsDefinition("http://xamarin.com/schemas/2020/toolkit", "Xamarin.CommunityToolkit.Converters")]
[assembly: XmlnsDefinition("http://xamarin.com/schemas/2020/toolkit", "Xamarin.CommunityToolkit.Effects")]
[assembly: XmlnsDefinition("http://xamarin.com/schemas/2020/toolkit", "Xamarin.CommunityToolkit.Extensions")]
[assembly: XmlnsDefinition("http://xamarin.com/schemas/2020/toolkit", "Xamarin.CommunityToolkit.UI.Views")]
```

File: XamarinCommunityToolkit/AssemblyInfo/AssemblyInfo.shared.cs [hunk 2]
```diff
@@ -1,3 +1,12 @@
...
-[assembly: InternalsVisibleTo("Xamarin.CommunityToolkit.UnitTests")]
\ No newline at end of file
+[assembly: InternalsVisibleTo("Xamarin.CommunityToolkit.UnitTests")]
+
+#if !NETSTANDARD1_0
+[assembly: XmlnsDefinition("http://xamarin.com/schemas/2020/toolkit", "Xamarin.CommunityToolkit.Behaviors")]
+[assembly: XmlnsDefinition("http://xamarin.com/schemas/2020/toolkit", "Xamarin.CommunityToolkit.Converters")]
+[assembly: XmlnsDefinition("http://xamarin.com/schemas/2020/toolkit", "Xamarin.CommunityToolkit.Effects")]
+[assembly: XmlnsDefinition("http://xamarin.com/schemas/2020/toolkit", "Xamarin.CommunityToolkit.Extensions")]
+[assembly: XmlnsDefinition("http://xamarin.com/schemas/2020/toolkit", "Xamarin.CommunityToolkit.UI.Views")]
```
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/AssemblyInfo/AssemblyInfo.shared.cs:11]: @jfversluis Why are **Views** inside **UI**, but other aren't?

**Comment by AndreiMisiukevich** [comment]
> Let's go with this one: http://xamarin.com/schemas/2020/toolkit
> 
> Just so we are inline with the rest and it looks super interesting 🤓

updated

File: XamarinCommunityToolkit/AssemblyInfo/AssemblyInfo.shared.cs (hunk 2 above)
Comment (jfversluis) [reply XamarinCommunityToolkit/AssemblyInfo/AssemblyInfo.shared.cs:11]: Because they are not visible, I guess 😅

Comment (AndreiMisiukevich) [reply XamarinCommunityToolkit/AssemblyInfo/AssemblyInfo.shared.cs:11]: Should we make all namespaces consistent? (With **UI** or w/out)

Comment (dansiegel) [reply XamarinCommunityToolkit/AssemblyInfo/AssemblyInfo.shared.cs:11]: if anything I would say drop the views... everything in UI.Views is a VisualElement... Behaviors, Converters, Effects, Extensions are all helpers but none of them are Visual Elements

**Comment by jfversluis** [comment]
Whooo boy! That are some conflicts :D

Sorry, did some work on the sample app... Let me know if you have the time and motivation to fix it, else I'll pick it up :)

**Comment by dansiegel** [comment]
wow... that list sure increased from when I started rebasing... I'll take care of it

**Comment by dansiegel** [comment]
@jfversluis alright everything is rebased and should be good to go now... for the love of all things holy... please be sure to merge this before making more changes to the sample app, or at least get me some more of that delicious Belgian beer and chocolate 🤣

**Comment by jfversluis** [comment]
Will do both! :D 

Yeah I wanted to merge this actually when I noticed that my change made a bazillion conflicts with this one. But I'm keeping my eye on the build progress and merge this in when it all looks good.

Thanks for this!

**Comment by roubachof** [comment]
Guys, be aware that to be able to consume types from the custom namespace, compiler requires that there's a code reference from the assembly that consumes the types, to the assembly that defines the types.

It means that we need to create an Initializer, like `XamarinCommunityToolkit.Initializer.Initialize()`.

Source is here: https://docs.microsoft.com/en-us/xamarin/xamarin-forms/xaml/custom-namespace-schemas#consuming-a-custom-namespace-schema

I just want to warn you cause I ran into that issues with Shadows: https://github.com/roubachof/Sharpnado.Shadows/issues/24

**Comment by dansiegel** [comment]
@roubachof not necessarily... keep in mind that the Community Toolkit multi-targets so you have one output assembly for each target, where Sharpnado uses an older style one assembly per target plus a shared assembly. The result is if someone simply uses anything from the Toolkit such as the AsyncCommand, ObservableRangeCollection, etc it would have the same effect.

Also we could simply add according to that article. Xamarin.Forms itself does this, and ultimately there are a number of reasons why you shouldn't be trying to link out more than the SDK.

```cs
[assembly: Preserve]
```

**Comment by roubachof** [comment]
Multi-target doesn't seem to change anything to the described issue:
Well if someone only use `Behaviors`, `Views` or anything just Xaml based and the toolkit has no code initializer, the user will end up with the same issue...

**Comment by jfversluis** [comment]
Hm if that's the case, that's not ideal.

On the other hand, what happens now? We say we are linker safe, if we only have a reference in XAML that should mean that everything still works, right? Does that change when we implement this? 

I guess it's a matter of verifying

**Comment by roubachof** [comment]
This has nothing to do with the linker here, you will be still linker safe.
It's just a Xaml compiler thing: you need to have a code reference towards the assembly which has custom xml namespace.
Or else:

> Failure to include such a code reference will result in the XAML compiler being unable to locate the assembly containing the custom namespace schema types.

**Comment by dansiegel** [comment]
What this boils down to is... is there a reference to the Xamarin Community Toolkit in code or not... So let's say that we have:

```xml
<ContentPage xmlns="http://xamarin.com/schemas/2014/forms"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:xct="http://xamarin.com/schemas/2020/toolkit"
             x:Class="XmlnsTest1.MainPage">
  <Image Source="{xct:GravatarImage Email='johndoe@gmail.com'}" />
</ContentPage
```

If this is our only reference to the Toolkit it would get stripped out... now let's say that somewhere else in my code which could be within the same project or a referenced project... let's say that I have:

```cs
public class ViewAViewModel : ObservableObject
{
}
```

Well now I have a reference to a type in code that comes from the Xamarin Community Toolkit assembly and the linker doesn't pull it out before XamlC runs... We could however add a dummy class like:

```cs
public static class CommunityToolkit
{
    public static void Init()
    {
        // Intentionally left blank... stupid linker...
    }
}
```

That way if someone is only using things within XAML and never has a reference in code, there is a simple dummy method that would keep the linker from pulling the reference. This would be used like:

```cs
public partial class App : Application
{
    public App()
    {
        CommunityToolkit.Init();
    }
}
```

**Comment by pictos** [comment]
@dansiege our goal is to don't have this kind of `init` API @jfversluis, please correct me if I'm saying something wrong.

That `[assembly: Preserve]` seems good, we can try this first and see how it goes.

**Comment by dansiegel** [comment]
@pictos after testing I'm not sure that will have the desired effect. I don't think it's the Mono linker that's really the issue here. If there is zero code reference in C# then you lose the reference to the assembly. 

There is something we could do like I did for the Prism Container Extensions where we introduce an assembly attribute that comes from the toolkit. It's totally transparent but can work as enough of a reference to keep you from losing the assembly. My tests showed if I used something like ObservableObject or ObservableRangeCollection then no omit is needed. Hope that makes sense.

**Comment by pictos** [comment]
> There is something we could do like I did for the Prism Container Extensions where we introduce an assembly attribute that comes from the toolkit. The attribute would be added via a hidden build prop.

Is there a code or sample that we can take a look at?

**Comment by dansiegel** [comment]
> our goal is to don't have this kind of init API

@pictos again this assumes that there is an opposition to having an empty Init somewhere. In that case I would do it a little differently here than what I had done in the [Prism Container Extensions](https://github.com/dansiegel/Prism.Container.Extensions/blob/7.2.0/src/Prism.DryIoc.Extensions/build/Prism.DryIoc.Extensions.targets). A little bit of hindsight, experience and just different goals. Remember that the Toolkit is currently marked LinkerSafe so we probably don't want to mess with that. The goal here is to ensure just enough of a reference to keep the reference to the assembly itself alive for XamlC if someone only ever uses types from the toolkit in XAML and not in code. If they have ANY reference whatsoever in code it preserves the reference for XamlC to function.

In the Toolkit we could add this hidden attribute. Note that it's not actually a preserve attribute, we just want something simple to reference.
```cs
[EditorBrowsable(EditorBrowsableState.Never)]
[AttributeUsage(AttributeTargets.Assembly, AllowMultiple = false)]
public sealed class PreserveToolkitAttribute : Attribute
{
}
```

Then what we would do is add a build directory to the nuget that would have a targets file and a code file like:

```cs
// PreserveToolkit.cs
[assembly: Xamarin.CommunityToolkit.PreserveToolkit]
```

```xml
<Project>
  <ItemGroup>
    <Compile Include="$(MSBuildThisFileDirectory)PreserveToolkit.cs" />
  </ItemGroup>
</Project>
```

What this would do is transparently add the attribute to the consuming assembly which would ensure that it doesn't get removed before XamlC runs.

**Comment by pictos** [comment]
@dansiegel thank you for this info, today I learned one more thing! I'll take a look at that tomorrow since is very late here. I hope that we can me this work with this attribute.

**Comment by dansiegel** [comment]
@jfversluis tested the CI build and it works as expected. Install the NuGet, and the Xmlns works without any references in C# to the community toolkit. Just made a small tweak to the targets to prevent the PreserveXamarinCommunityToolkit.cs file from showing up as a linked file in the IDE. This is ready to be merged :)

**Comment by pictos** [comment]
@dansiegel I'll test this and review it ASAP.
Thanks for this.

File: XamarinCommunityToolkitSample/Pages/Views/AvatarViewPage.xaml [hunk 3]
```diff
@@ -22,7 +21,7 @@
                                      Orientation="Horizontal"
                                      Spacing="10">
 
-                            <views:AvatarView ColorTheme="{x:Static views:ColorTheme.Jungle}"
+                            <xct:AvatarView ColorTheme="{x:Static xct:ColorTheme.Jungle}"
```
Comment (pictos) [inline XamarinCommunityToolkitSample/Pages/Views/AvatarViewPage.xaml:24]: You need to change the `views` to `xct`
```xml
    <views:AvatarView.Source>
            <UriImageSource Uri="{Binding Source}" />
    </views:AvatarView.Source>
</views:AvatarView>

**Review by pictos — CHANGES_REQUESTED** [review]
The building is failing, the reason is in my comment bellow

File: XamarinCommunityToolkitSample/Pages/Views/AvatarViewPage.xaml (hunk 3 above)
Comment (dansiegel) [reply XamarinCommunityToolkitSample/Pages/Views/AvatarViewPage.xaml:24]: good catch... not sure how my local build didn't fail

**Review by pictos — APPROVED** [review]
Tested on my end (VSWindows - Android, iOS, UWP, WPF) and works fine. Please test on your end as well.

The UWP project, for some reason, shows blank pages when we try to navigate to any category (Views, Behaviors, etc) but isn't caused by this PR, it's reproducible in the main branch.

**Review by jfversluis — APPROVED** [review]

**Pull request merged**